### PR TITLE
Added support for complex torch.pinverse

### LINKED ISSUE #45385: torch.pinverse for complex tensors and general complex tensor support
## 🚀 Feature
The SVD based torch.pinverse which works with complex tensors would be great. 

## Motivation

For low-rank matrices, this is an absolute must. And of course we can use the torch.svd functionality and then write the pseudoinverse ourselves, however this would make life a bit simpler. 

## Edit

One overarching more issue I've come across is the lack of functionality in PyTorch for complex tensors. Although I would like a pinverse function, I can still take the pseudoinverse of a complex tensor via the SVD. But then I ran into the issue that matrix multiplication of complex tensors is not supported. So I wrote a function to compute complex matmul. Then I went to start training the network and was hit with the warning that complex tensor gradients aren't fully supported. I am running PyTorch 1.6.0 by the way. 

So I guess my question would be if these issues are on the radar of the PyTorch team? And have you found that others would like this functionality as well? I, for one, would greatly benefit from full complex tensor support! Thanks for reading!

cc @ezyang @anjali411 @dylanbespalko @vincentqb @vishwakftw @jianyuh @nikitaved @pearu

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds support for complex-valued input for `torch.pinverse`.
Fixed cuda SVD implementation to return singular values with real dtype.

Fixes #45385.

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -100,10 +100,11 @@ Tensor pinverse(const Tensor& self, double rcond) {
...
   }
   Tensor U, S, V;
-  std::tie(U, S, V) = self.svd();
+  std::tie(U, S, V) = self.conj().svd();
   Tensor max_val = at::narrow(S, /*dim=*/-1, /*start=*/0, /*length=*/1);
-  Tensor S_pseudoinv = at::where(S > rcond * max_val, S.reciprocal(), at::zeros({}, self.options()));
-  return at::matmul(V, at::matmul(S_pseudoinv.diag_embed(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1), U.transpose(-2, -1)));
+  Tensor S_pseudoinv = at::where(S > rcond * max_val, S.reciprocal(), at::zeros({}, S.options()));
+  Tensor S_pseudoinv_diag = S_pseudoinv.diag_embed(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).to(self.dtype());
+  return at::matmul(V, at::matmul(S_pseudoinv_diag, U.transpose(-2, -1)));
```
Comment (nikitaved) [inline aten/src/ATen/native/LinearAlgebra.cpp:107]: Hmm, could you please explain why just `U.transpose` and not `U.transpose.conj`?
The SVD decomposition of `self` is `self = U S V*`, so `pinv = V pseudo_reciprocal(S) U*`, as `V` and `U` are Hermitian.

Comment (nikitaved) [reply aten/src/ATen/native/LinearAlgebra.cpp:107]: Ok, I see that you run `svd` on the conjugated result. But it could be confusing with this order and requires inserting some comments, maybe running svd directly on `self` is easier to understand?

Comment (IvanYashchuk) [reply aten/src/ATen/native/LinearAlgebra.cpp:107]: When I look at it now I agree with you that calling the svd directly on `self` is more clear.
I will change it to
```cpp
std::tie(U, S, V) = self.svd();
...
return at::matmul(V.conj(), at::matmul(S_pseudoinv_diag, U.transpose(-2, -1).conj()));
```

Comment (nikitaved) [reply aten/src/ATen/native/LinearAlgebra.cpp:107]: Also, we can save on one matrix product and diag_embed by computing `U S V*` as ` (U * S) @ V*`.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -102,8 +102,9 @@ Tensor pinverse(const Tensor& self, double rcond) {
   Tensor U, S, V;
   std::tie(U, S, V) = self.svd();
   Tensor max_val = at::narrow(S, /*dim=*/-1, /*start=*/0, /*length=*/1);
-  Tensor S_pseudoinv = at::where(S > rcond * max_val, S.reciprocal(), at::zeros({}, self.options()));
-  return at::matmul(V, at::matmul(S_pseudoinv.diag_embed(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1), U.transpose(-2, -1)));
+  Tensor S_pseudoinv = at::where(S > rcond * max_val, S.reciprocal(), at::zeros({}, S.options())).to(self.dtype());
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:105]: why do we need the `.to(self.dtype())` here?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -102,8 +102,9 @@ Tensor pinverse(const Tensor& self, double rcond) {
...
+  // computes V.conj() * diag(S_pseudoinv) * U.T.conj()
+  return at::matmul(V.conj() * S_pseudoinv.unsqueeze(-2), U.transpose(-2, -1).conj());
```
Comment (anjali411) [inline aten/src/ATen/native/LinearAlgebra.cpp:107]: why did we update `S_pseudoinv.diag_embed(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1)` to `S_pseudoinv.unsqueeze(-2)` given they give tensors of different dimensions?

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu [hunk 4]
```diff
@@ -1499,7 +1499,7 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda(const Tensor& self, bool som
     }
 
     U_working_copy = same_stride_to(U_working_copy, self.options());
-    S_working_copy = same_stride_to(S_working_copy, self.options());
+    S_working_copy = same_stride_to(S_working_copy, S_working_copy.options().device(self.device()));
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:1502]: should this bug fix go in 1.7?
cc. @nataliakliushkina

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 3 above)
Comment (nikitaved) [reply aten/src/ATen/native/LinearAlgebra.cpp:107]: diag_embed implies matrix multiplication, we can replace matrix multiplication with the point-wise product. Recall that `S` is a batch of diagonal elements.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu (hunk 4 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:1502]: The bug is that singular values for complex inputs would be returned with complex dtype? It's for you to decide @anjali411. Either way, there should be an svd test that would check returned type.

Comment (ngimel) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:1502]: @IvanYashchuk can you please update svd documentation to mention that svd supports complex matrices, and to mention the return types of S, U, V.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 2 above)
Comment (ngimel) [reply aten/src/ATen/native/LinearAlgebra.cpp:105]: type promotion does not work for `torch.where`, so this won't work for cases where S is real and self is complex

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 5]
```diff
@@ -102,8 +102,9 @@ Tensor pinverse(const Tensor& self, double rcond) {
   Tensor U, S, V;
   std::tie(U, S, V) = self.svd();
   Tensor max_val = at::narrow(S, /*dim=*/-1, /*start=*/0, /*length=*/1);
-  Tensor S_pseudoinv = at::where(S > rcond * max_val, S.reciprocal(), at::zeros({}, self.options()));
-  return at::matmul(V, at::matmul(S_pseudoinv.diag_embed(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1), U.transpose(-2, -1)));
+  Tensor S_pseudoinv = at::where(S > rcond * max_val, S.reciprocal(), at::zeros({}, S.options())).to(self.dtype());
+  // computes V.conj() * diag(S_pseudoinv) * U.T.conj()
```
Comment (ngimel) [inline aten/src/ATen/native/LinearAlgebra.cpp:106]: @ instead of * will be helpful in this comment

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu (hunk 4 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:1502]: Yes, that info will be added together with the fix for issue #45821.

**Comment by mruberry** [comment]
The documentation still needs to be updated to reflect that pinverse supports complex inputs on the CPU. A brief statement, like that in the documentation of torch.sub (https://pytorch.org/docs/master/generated/torch.sub.html?highlight=sub#torch.sub), should do it. 

Sample support statement: 
"Supports float inputs and complex inputs. Complex inputs are only supported on the CPU."

**Review by anjali411 — APPROVED** [review]
lgtm thanks @IvanYashchuk can you also add autograd tests for `pinverse` in a follow-up PR?

**Comment by IvanYashchuk** [comment]
> lgtm thanks @IvanYashchuk can you also add autograd tests for `pinverse` in a follow-up PR?

Yes, I'll do that.
